pkg/services/store/entity/key.go

```diff
@@ -14,35 +14,56 @@ type Key struct {
 }
 
 func ParseKey(key string) (*Key, error) {
-	// /<group>/<resource>/<namespace>(/<name>(/<subresource>))
-	parts := strings.SplitN(key, "/", 6)
-	if len(parts) < 4 {
-		return nil, fmt.Errorf("invalid key (expecting at least 3 parts): %s", key)
+
+	// /<group>/<resource>
+	parts := strings.SplitN(key, "/", 4)
+	if len(parts) < 3 {
+		return nil, fmt.Errorf("invalid key (expecting at least 2 parts): %s", key)
 	}
 
 	if parts[0] != "" {
 		return nil, fmt.Errorf("invalid key (expecting leading slash): %s", key)
 	}
 
-	k := &Key{
-		Group:     parts[1],
-		Resource:  parts[2],
-		Namespace: parts[3],
-	}
+	if len(parts) > 3 {
+		// NOTE: no leading slash in the leftover segment
+		// <namespace>(/<name>(/<subresource>))
+		partsAfterResource := strings.SplitN(parts[3], "/", 3)
 
-	if len(parts) > 4 {
-		k.Name = parts[4]
-	}
+		if len(partsAfterResource) == 0 {
+			return nil, fmt.Errorf("invalid key (expecting at least 1 part in a namespaced operation): %s", parts[3])
+		}
 
-	if len(parts) > 5 {
-		k.Subresource = parts[5]
-	}
+		k := &Key{
+			Group:     parts[1],
+			Resource:  parts[2],
+			Namespace: partsAfterResource[0],
+		}
+
+		if len(partsAfterResource) > 1 {
+			k.Name = partsAfterResource[1]
+		}
+
+		if len(partsAfterResource) > 2 {
+			k.Subresource = partsAfterResource[2]
+		}
 
-	return k, nil
+		return k, nil
+	} else {
+		return &Key{
+			Group:    parts[1],
+			Resource: parts[2],
+		}, nil
+	}
 }
 
 func (k *Key) String() string {
-	s := "/" + k.Group + "/" + k.Resource + "/" + k.Namespace
+	s := "/" + k.Group + "/" + k.Resource
+
+	if len(k.Namespace) > 0 {
+		s += "/" + k.Namespace
+	}
```

Comment:
would it ever make sense that we have a name without a namespace? If not we should probably switch the logic around so that if namespace=="" we return, otherwise we add namespace and continue

Reply:
It will make sense for cluster scoped resources. We currently only have `apiregistration.k8s.io` `APIService` that meets this criteria. However, we are explicitly. using file storage until watch gets supported for unified.

Reply:
poking around, keyFunc is defined here https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1535-L1553 and that default implementation is the reason that the keys don't match the url format. I'm not sure why the k8s authors decided to make the key format ambiguous, but basically it isn't possible to properly parse a key without knowing in advance whether it is namespaced or not, which kind of sucks.

Not sure what the best path forward here is, one option would be to define our own KeyFunc and KeyRootFunc which would allow us to use a key format that is unambiguous.

Reply:
At least right now, the place we'd need to inject those seems to be in GetAPIGroupInfo for each group, but we could maybe break that out into a shared helper.

Reply:
Good catch. Yeah, the ambiguity is a showstopper.

Would something like below work:

```
# cluster-scoped example
/apis/group/v1/clusterscopedkind =>  prefix + "/-/" + whatever
# namespaced example
/apis/group/v1/namespaces/default/kind => prefix + "/" + ns + whatever
```
Where `-` acts as a special "namespace" value for no-namespace?

Reply:
Will need to make sure that if we use it across the board, including with file storage, that using `-` as a directory doesn't break on any OS including Windows.

Reply:
I must add though, the ambiguity of the keys doesn't extend to its segment length (where segments are the strings extracted from a "/" split) - which is always 1 shorter than namespace keys. Is there any reason why we can't live with the ambiguity of the key prefix?

Reply:
yeah having a consistent namespace placeholder for non-namespaced keys would work I think, could be any string that is a valid folder name and won't be a namespace name. I had naively assumed that the key was just the url and never really thought about it beyond that.

Reply:
So I gave this a closer look. The only cluster scoped Kind authored by us that we currently have is `features`, but it is using Legacy storage. So there isn't a consumer for this kind of custom KeyFunc yet.

A bigger problem arises though with regards to kube-aggregator whose `APIService` (also ClusterScoped) is the using an actual K8s Store. The issue is that it doesn't expose the functionality to override it's `etcd` based storage:
https://github.com/kubernetes/kube-aggregator/blob/master/pkg/apiserver/apiserver.go#L255-L257

Of course, just passing in a restOptionsGetter based on file/unified storage allows one to use the etcd based storage without having to fork their package and exposing a way of specifying the store.

Pertinent places in Grafana, which lead to calling their NewDelegate: 
https://github.com/grafana/grafana/blob/main/pkg/services/apiserver/service.go#L374
https://github.com/grafana/grafana/blob/main/pkg/services/apiserver/aggregator/aggregator.go#L85

Given all this, I am not sure, how much do we actually gain by making Keys unambiguous. Since, keys themselves aren't persisted anywhere and they are an in-memory construct, one that is always read from left-to-right. I am not sure what's gained by absolutely guaranteeing unambiguity especially when it comes at the cost of forking another package which we currently are able to get by without.

Assuming, by reading them from left to right, we can make our code handling unambiguous even if the keys themselves are ambiguous. We do that by prioritizing the handling based on Kind names. If unified storage itself is aware which Kind its dealing with, at that top level, a determination is pretty easily made as to whether a namespace part will occur or not.

/cc @toddtreece

Reply:
the problem with that is that unified storage is explicitly kind-agnostic, and has no way of knowing (or reason for caring) whether a specific kind is namespaced or not. One alternative may be to divorce the entity store key format from the k8s key format and translate between them in the k8s storage implementation, but I'll have to think some more about that.